src/safe_mpc/abstract.py

```diff
@@ -0,0 +1,322 @@
+import numpy as np
+import re
+from copy import deepcopy
+import scipy.linalg as lin
+from casadi import SX, vertcat, fmax, norm_2, Function
+from acados_template import AcadosModel, AcadosSim, AcadosSimSolver, AcadosOcp, AcadosOcpSolver
+import torch
+import torch.nn as nn
+
+
+class NeuralNetDIR(nn.Module):
+    def __init__(self, input_size, hidden_size, output_size):
+        super(NeuralNetDIR, self).__init__()
+        self.linear_relu_stack = nn.Sequential(
+            nn.Linear(input_size, hidden_size),
+            nn.ReLU(),
+            nn.Linear(hidden_size, hidden_size),
+            nn.ReLU(),
+            nn.Linear(hidden_size, output_size),
+            nn.ReLU(),
+        )
+
+    def forward(self, x):
+        out = self.linear_relu_stack(x)
+        return out
+
+
+class AbstractModel:
+    def __init__(self, params):
+        self.params = params
+        self.amodel = AcadosModel()
+        # Dummy dynamics (double integrator)
+        self.amodel.name = "double_integrator"
+        self.x = SX.sym("x")
+        self.x_dot = SX.sym("x_dot")
+        self.u = SX.sym("u")
+        self.f_expl = self.u
+        self.p = SX.sym("p")
+        self.addDynamicsModel(params)
+        self.amodel.f_expl_expr = self.f_expl
+        self.amodel.x = self.x
+        self.amodel.xdot = self.x_dot
+        self.amodel.u = self.u
+        self.amodel.p = self.p
+
+        self.nx = self.amodel.x.size()[0]
+        self.nu = self.amodel.u.size()[0]
+        self.ny = self.nx + self.nu
+        self.nq = int(self.nx / 2)
+        self.nv = self.nx - self.nq
+
+        # Joint limits
+        self.u_min = -params.u_max * np.ones(self.nu)
+        self.u_max = params.u_max * np.ones(self.nu)
+        self.x_min = np.hstack([params.q_min * np.ones(self.nq), -params.dq_max * np.ones(self.nq)])
+        self.x_max = np.hstack([params.q_max * np.ones(self.nq), params.dq_max * np.ones(self.nq)])
+
+        # NN model (viability constraint)
+        self.nn_model = None
+        self.nn_func = None
+
+    def addDynamicsModel(self, params):
+        pass
+
+    def checkStateConstraints(self, x):
+        return np.all((x >= self.x_min) & (x <= self.x_max))
+
+    def checkControlConstraints(self, u):
+        return np.all((u >= self.u_min) & (u <= self.u_max))
+
+    def checkRunningConstraints(self, x, u):
+        return self.checkStateConstraints(x) and self.checkControlConstraints(u)
+
+    def checkSafeConstraints(self, x):
+        return True if float(self.nn_func(x, self.params.alpha).toarray()) >= 0. else False
+
+    def setNNmodel(self):
+        device = torch.device('cuda')
+        model = NeuralNetDIR(self.nx, (self.nx - 1) * 100, 1).to(device)
+        model.load_state_dict(torch.load(self.params.NN_DIR + 'model_3dof_vboc', map_location=device))
+        mean = torch.load(self.params.NN_DIR + 'mean_3dof_vboc')
+        std = torch.load(self.params.NN_DIR + 'std_3dof_vboc')
```

Comment:
Having hard-coded names that refer to the 3dof model does not seem appropriate in this class